File: paimon-core/src/test/java/org/apache/paimon/table/system/TagsTableTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.table.system;

import org.apache.paimon.catalog.Identifier;
import org.apache.paimon.data.BinaryString;
import org.apache.paimon.data.GenericRow;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.data.Timestamp;
import org.apache.paimon.manifest.ManifestCommittable;
import org.apache.paimon.predicate.Predicate;
import org.apache.paimon.predicate.PredicateBuilder;
import org.apache.paimon.reader.RecordReader;
import org.apache.paimon.schema.Schema;
import org.apache.paimon.table.FileStoreTable;
import org.apache.paimon.table.TableTestBase;
import org.apache.paimon.table.sink.TableCommitImpl;
import org.apache.paimon.table.source.ReadBuilder;
import org.apache.paimon.tag.Tag;
import org.apache.paimon.types.DataTypes;
import org.apache.paimon.utils.DateTimeUtils;
import org.apache.paimon.utils.Pair;
import org.apache.paimon.utils.TagManager;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import static org.assertj.core.api.Assertions.assertThat;

/** Unit tests for {@link TagsTable}. */
class TagsTableTest extends TableTestBase {

    private static final String tableName = "MyTable";
    private FileStoreTable table;
    private TagsTable tagsTable;
    private TagManager tagManager;

    @BeforeEach
    void before() throws Exception {
        Identifier identifier = identifier(tableName);
        Schema schema =
                Schema.newBuilder()
                        .column("product_id", DataTypes.INT())
                        .column("price", DataTypes.INT())
                        .column("sales", DataTypes.INT())
                        .primaryKey("product_id")
                        .option("tag.automatic-creation", "watermark")
                        .option("tag.creation-period", "daily")
                        .option("tag.num-retained-max", "3")
                        .build();
        catalog.createTable(identifier, schema, true);
        table = (FileStoreTable) catalog.getTable(identifier);
        TableCommitImpl commit = table.newCommit(commitUser).ignoreEmptyCommit(false);
        commit.commit(
                new ManifestCommittable(
                        0,
                        Timestamp.fromLocalDateTime(LocalDateTime.parse("2023-07-18T12:00:01"))
                                .getMillisecond()));
        commit.commit(
                new ManifestCommittable(
                        1,
                        Timestamp.fromLocalDateTime(LocalDateTime.parse("2023-07-19T12:00:01"))
                                .getMillisecond()));
        tagsTable = (TagsTable) catalog.getTable(identifier(tableName + "$tags"));
        tagManager = table.store().newTagManager();
        table.createTag("many-tags-test");
    }

    @Test
    void testTagsTable() throws Exception {
        List<InternalRow> expectRow = getExpectedResult();
        List<InternalRow> result = read(tagsTable);
        assertThat(result).containsExactlyElementsOf(expectRow);
    }

    @Test
    void testReadWithTagNameEqualFilter() throws Exception {
        table.createTag("tag-a");
        table.createTag("tag-b");
        table.createTag("tag-c");

        PredicateBuilder builder = new PredicateBuilder(TagsTable.TABLE_TYPE);
        assertThat(readTagNames(builder.equal(0, BinaryString.fromString("tag-b"))))
                .containsExactly("tag-b");

        assertThat(readTagNames(builder.equal(0, BinaryString.fromString("missing")))).isEmpty();
    }

    @Test
    void testReadWithTagNameInFilter() throws Exception {
        table.createTag("tag-a");
        table.createTag("tag-b");
        table.createTag("tag-c");

        PredicateBuilder builder = new PredicateBuilder(TagsTable.TABLE_TYPE);
        assertThat(
                        readTagNames(
                                builder.in(
                                        0,
                                        Arrays.asList(
                                                (Object) BinaryString.fromString("tag-a"),
                                                BinaryString.fromString("tag-c")))))
                .containsExactlyInAnyOrder("tag-a", "tag-c");
    }

    @Test
    void testReadWithTagNameNotEqualFilter() throws Exception {
        table.createTag("tag-a");
        table.createTag("tag-b");
        table.createTag("tag-c");

        PredicateBuilder builder = new PredicateBuilder(TagsTable.TABLE_TYPE);
        List<String> rows = readTagNames(builder.notEqual(0, BinaryString.fromString("tag-b")));
        assertThat(rows).contains("tag-a", "tag-c");
        assertThat(rows).doesNotContain("tag-b");
    }

    @Test
    void testReadWithNonTagNameFieldFilter() throws Exception {
        table.createTag("tag-a");
        table.createTag("tag-b");

        PredicateBuilder builder = new PredicateBuilder(TagsTable.TABLE_TYPE);
        long maxSnapshotId =
                tagManager.tagObjects().stream().mapToLong(p -> p.getKey().id()).max().orElse(0L);
        assertThat(readTagNames(builder.greaterOrEqual(1, maxSnapshotId))).isNotEmpty();
        assertThat(readTagNames(builder.greaterThan(1, maxSnapshotId))).isEmpty();
    }

    @Test
    void testReadWithNullFilterReturnsAll() throws Exception {
        table.createTag("tag-a");
        table.createTag("tag-b");

        List<String> all =
                tagManager.tagObjects().stream()
                        .map(Pair::getValue)
                        .collect(java.util.stream.Collectors.toList());
        assertThat(readTagNames(null)).containsExactlyInAnyOrderElementsOf(all);
    }

    private List<String> readTagNames(Predicate predicate) throws IOException {
        ReadBuilder readBuilder = tagsTable.newReadBuilder();
        if (predicate != null) {
            readBuilder = readBuilder.withFilter(predicate);
        }
        List<String> names = new ArrayList<>();
        try (RecordReader<InternalRow> reader =
                readBuilder.newRead().createReader(readBuilder.newScan().plan())) {
            reader.forEachRemaining(row -> names.add(row.getString(0).toString()));
        }
        return names;
    }

    private List<InternalRow> getExpectedResult() {
        Map<String, InternalRow> tagToRows = new TreeMap<>();
        for (Pair<Tag, String> snapshot : tagManager.tagObjects()) {
            Tag tag = snapshot.getKey();
            String tagName = snapshot.getValue();
            tagToRows.put(
                    tagName,
                    GenericRow.of(
                            BinaryString.fromString(tagName),
                            tag.id(),
                            tag.schemaId(),
                            Timestamp.fromLocalDateTime(
                                    DateTimeUtils.toLocalDateTime(tag.timeMillis())),
                            tag.totalRecordCount(),
                            tag.getTagCreateTime() == null
                                    ? null
                                    : Timestamp.fromLocalDateTime(tag.getTagCreateTime()),
                            tag.getTagTimeRetained() == null
                                    ? null
                                    : BinaryString.fromString(
                                            tag.getTagTimeRetained().toString())));
        }

        List<InternalRow> internalRows = new ArrayList<>();
        for (Map.Entry<String, InternalRow> entry : tagToRows.entrySet()) {
            internalRows.add(entry.getValue());
        }
        return internalRows;
    }
}

File: paimon-python/pypaimon/tests/system/tags_table_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""End-to-end tests for the ``$tags`` system table."""

import datetime
import os
import shutil
import tempfile
import unittest

from pypaimon import CatalogFactory, Schema
from pypaimon.common.json_util import JSON
from pypaimon.schema.data_types import DataField
from pypaimon.snapshot.snapshot import Snapshot
from pypaimon.table.system.tags_table import TagsTable


_TAG_COMMIT_MS = 1779100000000


def _seed_snapshot(table):
    snapshot_dir = "{}/snapshot".format(table.table_path)
    table.file_io.mkdirs(snapshot_dir)
    snapshot = Snapshot(
        version=3,
        id=7,
        schema_id=0,
        base_manifest_list="base-7.avro",
        delta_manifest_list="delta-7.avro",
        total_record_count=42,
        delta_record_count=42,
        commit_user="seed",
        commit_identifier=7,
        commit_kind="APPEND",
        time_millis=_TAG_COMMIT_MS,
    )
    table.file_io.try_to_write_atomic(
        "{}/snapshot-7".format(snapshot_dir), JSON.to_json(snapshot))
    table.file_io.try_to_write_atomic(
        "{}/LATEST".format(snapshot_dir), "7")
    table.file_io.try_to_write_atomic(
        "{}/EARLIEST".format(snapshot_dir), "7")


def _read(table):
    rb = table.new_read_builder()
    return rb.new_read().to_arrow(rb.new_scan().plan().splits())


class TagsTableTest(unittest.TestCase):

    def setUp(self):
        self.tmp = tempfile.mkdtemp(prefix="tags_sys_")
        warehouse = os.path.join(self.tmp, "warehouse")
        self.catalog = CatalogFactory.create({"warehouse": warehouse})
        self.catalog.create_database("db", False)
        fields = [DataField.from_dict({"id": 0, "name": "v", "type": "INT"})]
        self.catalog.create_table("db.t", Schema(fields=fields), False)
        self.table = self.catalog.get_table("db.t")
        _seed_snapshot(self.table)

    def tearDown(self):
        shutil.rmtree(self.tmp, ignore_errors=True)

    def test_tags_table_loaded_via_catalog(self):
        table = self.catalog.get_table("db.t$tags")
        self.assertIsInstance(table, TagsTable)

    def test_schema_column_layout(self):
        table = self.catalog.get_table("db.t$tags")
        row_type = table.row_type()
        self.assertEqual(
            ["tag_name", "snapshot_id", "schema_id", "commit_time",
             "record_count", "create_time", "time_retained"],
            [f.name for f in row_type.fields],
        )
        expected_nullability = [False, False, False, False, True, True, True]
        for field, expected in zip(row_type.fields, expected_nullability):
            self.assertEqual(expected, field.type.nullable,
                             "field {} nullability".format(field.name))
        self.assertEqual(["tag_name"], table.primary_keys())

    def test_empty_when_no_tags_created(self):
        arrow_table = _read(self.catalog.get_table("db.t$tags"))
        self.assertEqual(0, arrow_table.num_rows)

    def test_lists_created_tags_with_snapshot_metadata(self):
        self.table.create_tag("v1", snapshot_id=7)
        self.table.create_tag("v2", snapshot_id=7)

        arrow_table = _read(self.catalog.get_table("db.t$tags"))
        names = arrow_table.column("tag_name").to_pylist()
        self.assertEqual({"v1", "v2"}, set(names))

        snapshot_ids = arrow_table.column("snapshot_id").to_pylist()
        schema_ids = arrow_table.column("schema_id").to_pylist()
        record_counts = arrow_table.column("record_count").to_pylist()
        self.assertEqual([7, 7], snapshot_ids)
        self.assertEqual([0, 0], schema_ids)
        self.assertEqual([42, 42], record_counts)

        for ts in arrow_table.column("commit_time").to_pylist():
            self.assertIsInstance(ts, datetime.datetime)
            ms = int(ts.replace(
                tzinfo=datetime.timezone.utc).timestamp() * 1000)
            self.assertEqual(_TAG_COMMIT_MS, ms)

        # Until pypaimon Tag carries these fields they are surfaced as
        # None — same trade-off as FileSystemCatalog.get_tag.
        for value in arrow_table.column("create_time").to_pylist():
            self.assertIsNone(value)
        for value in arrow_table.column("time_retained").to_pylist():
            self.assertIsNone(value)


if __name__ == "__main__":
    unittest.main()
